nncf/quantization/waveq_loss.py

```diff
@@ -0,0 +1,76 @@
+import math
+import torch
+
+from nncf.compression_method_api import CompressionLoss
+from nncf.quantization.algo import BaseQuantizer
+from nncf.quantization.layers import SymmetricQuantizer, AsymmetricQuantizer
+
+
+class LossHook:
+
+    def __init__(self, quant_module: BaseQuantizer):
+        self.input_tensor = None
+        self.out_tensor = None
+        self.quant_module = quant_module
+
+    def fill_post_hook(self, module, inputs=None, outputs=None):
+        self.input_tensor = inputs[0]
+        self.out_tensor = outputs
+
+
+class WaveQLoss(CompressionLoss):
+    # change to weights quant cnly
+    def __init__(self, quantize_modules, ratio=1):
+        super().__init__()
+        self.ratio = ratio
+        self.quantize_modules = quantize_modules
+        self.post_hook_handlers = None
+        self.pre_hook_handlers = None
+        self.hooks = None
+        self.set_up_hooks()
+
+    def set_up_hooks(self):
+        self.post_hook_handlers = []
+        self.hooks = []
+        for module in self.quantize_modules:
+            hook = LossHook(module)
+            self.post_hook_handlers.append(module.register_forward_hook(hook.fill_post_hook))
+            self.hooks.append(hook)
+
+    def forward(self):
+        loss = 0
+        for hooker in self.hooks:
+            loss += WaveQLoss.waveq_loss_per_hook_sum(hooker, ratio=self.ratio)
+        return loss
+
+    def get_hook_data(self):
+        #get input and quant_module
+        pass
+
+    @staticmethod
+    def waveq_loss_for_tensor(tensor: torch.tensor, ratio=1, levels=16, input_low=0, input_range=1):
+        # device problem
+        # check device
+        return ratio * torch.square(torch.sin((tensor + input_low) / input_range
+                                              * (levels - 1) * math.pi)) / levels
+
+    @staticmethod
+    def waveq_loss_per_hook_sum(hook: LossHook, ratio=1):
+        input_low, input_range, levels = get_quant_module_params(hook.quant_module)
+        out = WaveQLoss.waveq_loss_for_tensor(hook.input_tensor, ratio,
+                                              levels=levels, input_low=input_low, input_range=input_range)
+        return torch.sum(out)
+
+
+def get_quant_module_params(quant_module: BaseQuantizer):
+    try:
+        quant_module.signed
```

Comment:
what if exception happens later? it's wrong logic. quick & dirty fix is to use `isinstance()` method. 
But I would add to `BaseQuantizer` `get_quantization_ranges` method and return final input_low and input_range. 

Please pay attention to some modifications of parameters:
for symmetric
```
scale_safe = abs(scale) + eps
```
and assymetric
```
    input_range_safe = abs(input_range) + eps
    input_low_tuned, input_range_tuned = TuneRange.apply(input_low, input_range_safe, levels)
```

Without these modifications WaveQ loss can be inaccurate.